tests/models/test_hooks.py

```diff
@@ -872,20 +878,12 @@ def call(hook, fn, *args, **kwargs):
     called = []
     dm = HookedDataModule(called)
     trainer.fit(model, datamodule=dm)
-    batch_transfer = [
-        dict(name="on_before_batch_transfer", args=(ANY, 0)),
-        dict(name="transfer_batch_to_device", args=(ANY, torch.device("cpu"), 0)),
-        dict(name="on_after_batch_transfer", args=(ANY, 0)),
-    ]
```

Comment:
This test only aims to test the order for the `LightningDataModule` hooks, however, the transfer of batches is part of the LightningModule. Why were these here then?

The reason is that `LightningDataModule` subclasses `DataHooks`.
https://github.com/PyTorchLightning/pytorch-lightning/blob/ba036fdeeade11aeb43f3f35ff1b924b3780adc1/pytorch_lightning/core/datamodule.py#L26
And `DataHooks` implements the transfer hooks.
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/hooks.py#L709
So for this test both the `HookedDataModule` and `BoringModel` share the same implementation (none of them override it).

This means that when we were wrapping it with partial above, we were doing it for this shared implementation and thus the hook was getting included in `called` for the DataModule when in reality it was for the `LightningModule`.

`update_wrapper` fixes that.

cc @awaelchli

Reply:
Thanks, that's pretty nasty.
Do you think a testing approach with unittest.Mock would have helped to prevent this here? (aware we are currently limited by some internals preventing to use Mock)

Reply:
I wouldn't know without trying